server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -0,0 +1,491 @@
+                                        }
+
+                                        @Override
+                                        public String toString() {
+                                            return "TransportAddressConnector connecting to " + clusterNode.getId();
+                                        }
+                                    });
+                                    return;
+                                }
+                            }
+                            deterministicTaskQueue.scheduleNow(new Runnable() {
+                                @Override
+                                public void run() {
+                                    listener.onFailure(
+                                        new ElasticsearchException("no such node: " + transportAddress + " in " + clusterNodes));
+                                }
+
+                                @Override
+                                public String toString() {
+                                    return "TransportAddressConnector failing to connect to " + transportAddress;
+                                }
+                            });
+                        };
+                    }
+
+                    @Override
+                    protected void connectToNode(DiscoveryNode node) {
+                        // do nothing, we are simulating connections between nodes.
+                    }
+                };
+
+                coordinator.start();
+                coordinator.startInitialJoin();
+            }
+
+            private PersistedState getPersistedState() {
+                return persistedState;
+            }
+
+            String getId() {
+                return localNode.getId();
+            }
+
+            public DiscoveryNode getLocalNode() {
+                return localNode;
+            }
+
+            boolean isLeader() {
+                return coordinator.getMode() == Coordinator.Mode.LEADER;
+            }
+
+            void submitValue(final long value) {
+                masterService.submitStateUpdateTask("new value [" + value + "]", new ClusterStateUpdateTask() {
+                    @Override
+                    public ClusterState execute(ClusterState currentState) {
+                        return setValue(currentState, value);
+                    }
+
+                    @Override
+                    public void onFailure(String source, Exception e) {
+                        logger.debug(() -> new ParameterizedMessage("failed to publish: [{}]", source), e);
+                    }
+                });
+            }
+        }
+
+        private List<TransportAddress> provideUnicastHosts(HostsResolver ignored) {
+            final List<TransportAddress> unicastHosts = new ArrayList<>(clusterNodes.size());
+            clusterNodes.forEach(n -> unicastHosts.add(n.getLocalNode().getAddress()));
+            return unmodifiableList(unicastHosts);
```

Comment:
This would look better with stream():

```
return clusterNodes.stream().map(ClusterNode::getLocalNode).map(DiscoveryNode::getAddress).collect(Collectors.toList());
```